### PR TITLE
Add support for implicit load/store operations

### PR DESCRIPTION
**Pull request by kanishkan**
Proposed auto-increment load/store ISAs extensions

### CODE REVIEW

**Comment by kanishkan** [comment]
We do not have hardware backend for these instructions. So I am not sure if it makes sense to add these ISA in document yet.
@pjaaskel : any thoughts?

File: tce/opset/base/Makefile.am [hunk 1]
```diff
@@ -2,18 +2,21 @@ basedir = ${prefix}/share/tce/opset/base
 avalondir = ${prefix}/share/tce/opset/base
 simple_iodir = ${prefix}/share/tce/opset/base
 doubledir = ${prefix}/share/tce/opset/base
+implicit_lsudir = ${prefix}/share/tce/opset/base
```
Comment (pjaaskel) [inline tce/opset/base/Makefile.am:5]: Strictly put LSU is for load-store unit. Here you are defining load-store _operations_ (that can be in function units). Implicit is also too generic word, what you are doing is postincrement_load_store, right?

File: tce/opset/base/implicit_lsu.cc [hunk 2]
```diff
@@ -0,0 +1,300 @@
+/*
+    Copyright (c) 2002-2020 Tampere University.
```
Comment (pjaaskel) [inline tce/opset/base/implicit_lsu.cc:2]: (c) belongs to you or Eindhoven University. We have only agreed on licensing, not copyright transfer.

File: tce/opset/base/implicit_lsu.cc [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+ * Implicit LSU operation definitions.
+ *
+ * @author Kanishkan Vadivel (k.vadivel-no.spam-tue.nl)
+ * @author Barry de Bruin (e.d.bruin-no.spam-tue.nl)
+ */
+
+#include "OSAL.hh"
+#include "OperationGlobals.hh"
+
+#define RUNTIME_ERROR_WITH_INT(MESSAGE, DATA) {\
```
Comment (pjaaskel) [inline tce/opset/base/implicit_lsu.cc:36]: Can you make a generic RUNTIME_ERROR with a varlength so we do not need this special macros for different parameters?

File: tce/opset/base/Makefile.am (hunk 1 above)
Comment (pjaaskel) [reply tce/opset/base/Makefile.am:5]: But perhaps we want to add postdecrement too in the future so "implicit_load_store" or such?

File: tce/opset/base/implicit_lsu.cc [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+//////////////////////////////////////////////////////////////////////////////
+// set LSU configuration register
+//////////////////////////////////////////////////////////////////////////////
+OPERATION_WITH_STATE(ILD_ADDR, IMPLICIT_LSU_STATE)
+TRIGGER
+    STATE.loadAddr = UINT(1);
+END_TRIGGER;
+END_OPERATION_WITH_STATE(ILD_ADDR)
+
+OPERATION_WITH_STATE(ILD_STRIDE, IMPLICIT_LSU_STATE)
```
Comment (pjaaskel) [inline tce/opset/base/implicit_lsu.cc:75]: Could we combine these to one operation to reduce needed opcodes to support the implicit load/stores? ILD_INIT or such?

File: tce/opset/base/implicit_lsu.cc [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+END_OPERATION_WITH_STATE(IST_ADDR)
+
+OPERATION_WITH_STATE(IST_STRIDE, IMPLICIT_LSU_STATE)
+TRIGGER
+    STATE.storeStride = UINT(1);
+END_TRIGGER;
+END_OPERATION_WITH_STATE(IST_STRIDE)
+
+//////////////////////////////////////////////////////////////////////////////
+// LD8 - load a 8-bit little endian word from memory with sign extension
```
Comment (pjaaskel) [inline tce/opset/base/implicit_lsu.cc:94]: Endianness concerns only multi byte words.

File: testsuite/systemtest/bintools/Compiler/data/implicit_ldst.c [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        _TCEAS_ILD_ADDR("#0", &input_samples[i]);
+        _TCEAS_ILD_ADDR("#1", &coeffs[0]);
+
+        #pragma clang loop unroll(disable)
+        for (int c = 0; c < 8; c++) {
+            _TCEAS_ILD32("#1", UNUSED, weight);
+            _TCEAS_ILD32("#0", UNUSED, data);
+            sum += weight * data;
+        }
+        _TCEAS_IST32("#0", sum);
```
Comment (pjaaskel) [inline testsuite/systemtest/bintools/Compiler/data/implicit_ldst.c:73]: Can you add a test for the load-store combo operations?

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Thanks. Very useful feature. A couple of small things to consider only. I'd try to keep the opcode count low as they reflect to the instruction width.

File: tce/opset/base/implicit_lsu.cc (hunk 3 above)
Comment (kanishkan) [reply tce/opset/base/implicit_lsu.cc:36]: moved to single macro.

File: tce/opset/base/implicit_lsu.cc (hunk 2 above)
Comment (kanishkan) [reply tce/opset/base/implicit_lsu.cc:2]: Fixed

File: tce/opset/base/implicit_lsu.cc (hunk 5 above)
Comment (kanishkan) [reply tce/opset/base/implicit_lsu.cc:94]: yes indeed.

File: tce/opset/base/implicit_lsu.cc (hunk 4 above)
Comment (kanishkan) [reply tce/opset/base/implicit_lsu.cc:75]: Moved to ILD_INIT and IST_INIT

File: tce/opset/base/Makefile.am (hunk 1 above)
Comment (kanishkan) [reply tce/opset/base/Makefile.am:5]: sure

Comment (kanishkan) [reply tce/opset/base/Makefile.am]: fixed

File: testsuite/systemtest/bintools/Compiler/data/implicit_ldst.c (hunk 6 above)
Comment (kanishkan) [reply testsuite/systemtest/bintools/Compiler/data/implicit_ldst.c:79]: added.

**Comment by kanishkan** [comment]
@pjaaskel : Updated the patch for your comments. 
In terms of instructions, I just combined `ILD/IST_STRIDE` and `ILD/IST_ADDR` to `ILD/IST_INIT` instruction.

**Comment by kanishkan** [comment]
Also testsuite is passing.

File: tce/opset/base/implicit_load_store.cc [hunk 7]
```diff
@@ -0,0 +1,283 @@
+/*
+    Copyright (c) 2002-2021 Eindhoven University of Technology
```
Comment (pjaaskel) [inline tce/opset/base/implicit_load_store.cc:2]: Nitpick: They year you authored this is only limited to 2021, right?

File: tce/opset/base/implicit_load_store.cc [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+ * Implicit LSU operation definitions.
+ *
+ * @author Kanishkan Vadivel (k.vadivel-no.spam-tue.nl)
+ * @author Barry de Bruin (e.d.bruin-no.spam-tue.nl)
+ */
+
+#include "OSAL.hh"
+#include "OperationGlobals.hh"
+
+#define RUNTIME_ERROR_WITH_LONG(MESSAGE, DATA) {\
```
Comment (pjaaskel) [inline tce/opset/base/implicit_load_store.cc:36]: I mean to implement using a vararg macro: "RUNTIME_ERROR(FMT, ...)"

**Review by pjaaskel — APPROVED** [review]

**Pull request merged**